### PR TITLE
feat: Add pinned memory optimizer offload for Megatron policy worker

### PR DESCRIPTION
**Pull request by snivertynv**
This significantly improves performance for the optimizer_offload_before_refit pass which is quite expensive in co-located/syncRL cases. 

Enabled/disabled using the `use_pinned_optimizer_offload` setting (default=disabled). It has been set to `false` in a couple of grpo_math* yaml configs as an example. Added test cases for this feature in `test_megatron_worker.py`

# What does this PR do ?

Optimizer D2H/H2D transfers used per-tensor pageable allocations, causing expensive cudaHostAlloc calls and synchronous memcpy on every step. This adds an opt-in mode (use_pinned_optimizer_offload) that coalesces all optimizer state into a single cached pinned buffer, eliminating cudaHostAlloc from the hot path and enabling non-blocking DMA transfers.

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

**Review by terrykong — COMMENTED** [review]
@guyueh1 could you help review?

File: nemo_rl/models/policy/workers/megatron_policy_worker.py [hunk 1]
```diff
@@ -1255,22 +1255,92 @@ def move_optimizer(self, device: str):
             optimizer_state = self.optimizer.state
         else:
             optimizer_state = self.optimizer._get_state()
+
+        use_pinned = self.cfg.get("use_pinned_optimizer_offload", False)
```
Comment (guyueh1) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py:1259]: this is not ideal, we should use `self.cfg['use_pinned_optimizer_offload']`; we should add this field to all necessary base configs under `examples/configs` to make sure the read doesn't fail.

File: nemo_rl/models/policy/workers/megatron_policy_worker.py [hunk 2]
```diff
@@ -1255,22 +1255,92 @@ def move_optimizer(self, device: str):
...
+        """Reload optimizer state tensors to CUDA."""
+        for _, state in optimizer_state.items():
+            for k, v in state.items():
+                if torch.is_tensor(v) and not v.is_cuda:
+                    state[k] = v.to("cuda")
+
+    def _get_or_alloc_pinned_buf(
+        self, attr_name: str, total_bytes: int
+    ) -> torch.Tensor:
+        """Return a cached pinned CPU buffer, allocating only on first use or resize."""
```
Comment (guyueh1) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py:1293]: I recommend having a `def _delete_pinned_buf(self, attr_name)` paired with it; in the optimizer offload case, we repeatedly use the same buffer and we don't want to delete until job finish, but if you leave this API here and someone implements some new temporal offloading with it, they will assume you handled the deletion but actually not. 

So either we add a deletion api here too and show people how to use them in pairs, or we make this allocation API specific to optimizer offloading (in naming or just hardcode it in the _optimizer_to functions), so that folks don't use it for other purposes.

File: nemo_rl/models/policy/workers/megatron_policy_worker.py (hunk 1 above)
Comment (snivertynv) [reply nemo_rl/models/policy/workers/megatron_policy_worker.py]: Fixed.

File: nemo_rl/models/policy/workers/megatron_policy_worker.py (hunk 2 above)
Comment (snivertynv) [reply nemo_rl/models/policy/workers/megatron_policy_worker.py:2255]: Thanks for this feedback, fixed it.

**Comment by guyueh1** [comment]
I recently found an issue with this implementation, when I am testing a deepseek model on Gb200 using this PR, it throws CPU OOM because we always allocate a big contiguous buffer, when I changed it to this fine-grained pinned memory allocation method ([commit](https://github.com/NVIDIA-NeMo/RL/commit/8a13d5ad7e4a1e772a0d0dadcc08b8821de9be9e)) it worked for me. 

This makes me think, if we should separate pinned memory and coalesced buffer as two features that can be turned on/off separated. It may not make sense to use unpinned but coalesced, but there are certainly scenarios where I need pinned and un-coalesced. @snivertynv what do you think?

**Review by guyueh1 — DISMISSED** [review]

**Review by guyueh1 — DISMISSED** [review]

**Review by terrykong — DISMISSED** [review]

File: docs/broken_links_needing_review.json [hunk 3]
```diff
@@ -0,0 +1,2 @@
+{"uri": "https://docs.pytorch.org/docs/stable/notes/cuda.html#optimizing-memory-usage-with-pytorch-alloc-conf"}
```
Comment (guyueh1) [inline docs/broken_links_needing_review.json]: why this? where does these link appear?

**Review by guyueh1 — COMMENTED** [review]
LGTM except i'm confused by the added broken_links_needing_review.json. The two links seem to work in browser (redirected to a pytorch 2.12 doc)

File: docs/broken_links_needing_review.json (hunk 3 above)
Comment (snivertynv) [reply docs/broken_links_needing_review.json]: Hi @guyueh1 , these links appear in `docs/guides/dtensor-tp-accuracy.md` and `docs/about/tips-and-tricks.md`. They're valid URLs that redirect in a browser, but the Sphinx linkcheck CI job reports them as broken because it doesn't follow redirects to re-verify anchors (PyTorch docs redirect `stable/` to `2.12/`). So we tell the exclusion file and therefore the linkcheck to skip these so CI doesn't fail on these links which are still technically correct. But if you think this isn't needed, I can remove.

Comment (guyueh1) [reply docs/broken_links_needing_review.json]: how does CI know to parse this file to look for skippable URLs for lint check? Does ci recognize this file's name?

**Review by guyueh1 — DISMISSED** [review]

File: examples/configs/grpo_math_1B.yaml [hunk 4]
```diff
@@ -111,6 +111,8 @@ policy:
   precision: "bfloat16"
   logprob_chunk_size: null
   offload_optimizer_for_logprob: false # Only useful for non-colocated generation since colocated generation will always offload optimizer to cuda before refit
+  use_pinned_optimizer_offload: false # Use pinned memory for optimizer D2H/H2D transfers (per-tensor pinning)
+  use_coalesced_optimizer_offload: false # Coalesce all optimizer state into a single pinned buffer; requires use_pinned_optimizer_offload=true. Faster but may OOM on very large models.
```
Comment (terrykong) [inline examples/configs/grpo_math_1B.yaml]: [`grpo_math_1B.yaml:114-115`](https://github.com/NVIDIA-NeMo/RL/blob/17f931c1b8aa1abd8fb3e781e4ff44c3ffa71e57/examples/configs/grpo_math_1B.yaml#L114-L115)

The PR description doesn't give much guidance on *when* a user should enable these two flags. Could you expand the inline comments (and ideally a line in the PR description) to spell out the scenarios each one helps in — and, since this is a Performance change, roughly what speedup to expect? Something like:

```suggestion
  use_pinned_optimizer_offload: false # Pin optimizer state in CPU memory for non-blocking D2H/H2D transfers. Enable for colocated/syncRL runs that offload the optimizer to CPU each step; no benefit if the optimizer is never offloaded. Uses extra pinned host memory.
  use_coalesced_optimizer_offload: false # Pack all optimizer state into one cached pinned buffer (requires use_pinned_optimizer_offload=true). Enable when per-tensor cudaHostAlloc overhead dominates offload time; leave false (plain pinning) if the single contiguous buffer risks CPU OOM on very large / MoE models.
```

File: nemo_rl/models/policy/__init__.py [hunk 5]
```diff
@@ -425,6 +425,9 @@ class PolicyConfig(TypedDict):
     sequence_packing: NotRequired[SequencePackingConfig | SequencePackingConfigDisabled]
     make_sequence_length_divisible_by: int
     max_total_sequence_length: int
+    offload_optimizer_for_logprob: bool
+    use_pinned_optimizer_offload: bool
+    use_coalesced_optimizer_offload: bool
```
Comment (terrykong) [inline nemo_rl/models/policy/__init__.py:549]: [`policy/__init__.py:429-430`](https://github.com/NVIDIA-NeMo/RL/blob/17f931c1b8aa1abd8fb3e781e4ff44c3ffa71e57/nemo_rl/models/policy/__init__.py#L429-L430)

It would help to document when to enable these right where they're defined. Suggested doc comments:

```suggestion
    # Pin optimizer state in CPU memory so optimizer D2H/H2D offload transfers are non-blocking.
    # Enable for colocated/syncRL runs that offload the optimizer to CPU each step; no effect if the
    # optimizer is never offloaded. Costs extra pinned host memory.
    use_pinned_optimizer_offload: bool
    # Pack all optimizer state into one cached pinned buffer. Requires use_pinned_optimizer_offload=True
    # (enforced at runtime in move_optimizer). Enable on top of pinning when per-tensor cudaHostAlloc
    # overhead dominates offload time; prefer plain pinning if the single contiguous buffer risks CPU
    # OOM on very large / MoE models.
    use_coalesced_optimizer_offload: bool
```

Separately, on the dependency between them: the `use_coalesced ⇒ use_pinned` requirement is only enforced at runtime inside [`move_optimizer()`](https://github.com/NVIDIA-NeMo/RL/blob/17f931c1b8aa1abd8fb3e781e4ff44c3ffa71e57/nemo_rl/models/policy/workers/megatron_policy_worker.py#L1586-L1589) (`raise ValueError(...)`), which doesn't fire until the first optimizer offload — potentially deep into a run. Consider validating it at config/worker-init time so a misconfiguration (`coalesced=true, pinned=false`) fails fast at startup instead of mid-training.

File: docs/broken_links_needing_review.json (hunk 3 above)
Comment (terrykong) [reply docs/broken_links_needing_review.json]: should be fixed in https://github.com/NVIDIA-NeMo/RL/pull/2889/changes. once you rebase, you can drop this change

File: nemo_rl/models/policy/workers/megatron_policy_worker.py [hunk 6]
```diff
@@ -1503,8 +1503,20 @@ def offload_after_refit(self):
...
+            self.model, "cpu", move_params=False, move_grads=True
+        )
         torch.randn(1).cuda()  # wake up torch allocator
-        self.offload_before_refit()  # rerun the old offload function
+        if (
+            hasattr(self, "optimizer")
+            and self.optimizer is not None
+            and not self.optimizer_cpu_offload
+        ):
+            self.move_optimizer("cpu")
```
Comment (terrykong) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py:2117]: [`megatron_policy_worker.py:1516`](https://github.com/NVIDIA-NeMo/RL/blob/17f931c1b8aa1abd8fb3e781e4ff44c3ffa71e57/nemo_rl/models/policy/workers/megatron_policy_worker.py#L1516)

A note on the refactor of `offload_after_refit`. Previously this method ended by calling [`self.offload_before_refit()`](https://github.com/NVIDIA-NeMo/RL/blob/17f931c1b8aa1abd8fb3e781e4ff44c3ffa71e57/nemo_rl/models/policy/workers/megatron_policy_worker.py#L1409), reusing everything that function does. This PR instead inlines those steps directly here — move grad buffers, then `move_optimizer("cpu")`, then `gc.collect()` / `empty_cache()`.

The inlined version reproduces the offload steps, but it drops the two **optional** cache-clearing blocks that `offload_before_refit()` still runs ([L1422-L1478](https://github.com/NVIDIA-NeMo/RL/blob/17f931c1b8aa1abd8fb3e781e4ff44c3ffa71e57/nemo_rl/models/policy/workers/megatron_policy_worker.py#L1422-L1478)):

- `_clear_fp8_caches()` — when `fp8_cfg.force_clear_fp8_caches` is set
- the RotaryEmbedding `lru_cache` clear + MoE token-dispatcher tensor nulling — when `megatron_cfg.clear_memory_caches_before_refit` is set

Both flags default to `false`, so no in-repo config is affected today. But for a user who enables either one (e.g. a MoE/fp8 run), the caches that `offload_before_refit` used to free are no longer freed in `offload_after_refit` — so GPU memory can creep across a run. Not a correctness bug.

Was inlining (and leaving out those two blocks) intentional? If the goal was just to make `offload_after_refit` self-contained, it'd be safer to either call `offload_before_refit()` again or factor the shared offload + cache-clearing into a single helper that both methods call, so the two paths can't drift apart.

Comment (terrykong) [reply nemo_rl/models/policy/workers/megatron_policy_worker.py:2117]: cc @ZhiyuLi-Nvidia to opine here. also @ZhiyuLi-Nvidia can you suggest a test that might be good to catch such a memory regression for fp8? this seems easy to miss from code-review, better to have something catch it

File: nemo_rl/models/policy/workers/megatron_policy_worker.py [hunk 7]
```diff
@@ -1503,8 +1503,20 @@ def offload_after_refit(self):
         no_grad.__enter__()
         self.model = self.move_model(self.model, "cpu")
         self.model.eval()
+        # Offload grad buffers (params already on CPU from the move above).
+        self.model = self.move_model(
+            self.model, "cpu", move_params=False, move_grads=True
+        )
```
Comment (terrykong) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py]: [`megatron_policy_worker.py:1504-1509`](https://github.com/NVIDIA-NeMo/RL/blob/17f931c1b8aa1abd8fb3e781e4ff44c3ffa71e57/nemo_rl/models/policy/workers/megatron_policy_worker.py#L1504-L1509)

This can be collapsed to a single `move_model`. `move_model` defaults to `move_params=True, move_grads=True` ([def](https://github.com/NVIDIA-NeMo/RL/blob/17f931c1b8aa1abd8fb3e781e4ff44c3ffa71e57/nemo_rl/models/policy/workers/megatron_policy_worker.py#L1529)), so the first call on line 1504 already offloads **both** params and grad buffers to CPU. That makes the second `move_model(self.model, "cpu", move_params=False, move_grads=True)` redundant — the grad buffers it moves were already moved by line 1504 (the `# params already on CPU from the move above` comment is right, but the same call also offloaded the grads). Dropping it:

```suggestion
        self.model = self.move_model(self.model, "cpu")
        self.model.eval()
```

**Review by terrykong — COMMENTED** [review]
Thanks for this — the opt-in pinned/coalesced optimizer offload is a coherent, well-tested feature, and the core offload/reload logic checks out (round-trip value/dtype preservation across pageable/pinned/coalesced, buffer caching/growth, and the `coalesced⇒pinned` invariant are all covered and pass).

A few supplementary notes below in addition to @guyueh1's approval.

**Merge state:** the PR currently shows `CONFLICTING` / `DIRTY` — it needs a rebase on `main` and conflict resolution before it can merge. (The submodule fast-forward check posted earlier in the thread is now stale; it and Lint/copyright/sphinx-build all pass on the latest commit.)

_Generated by Claude Code (6-agent team review)_

File: docs/broken_links_needing_review.json (hunk 3 above)
Comment (snivertynv) [reply docs/broken_links_needing_review.json]: Dropped from latest commit, thanks

File: nemo_rl/models/policy/workers/megatron_policy_worker.py (hunk 6 above)
Comment (terrykong) [reply nemo_rl/models/policy/workers/megatron_policy_worker.py:2117]: fixed in 16ccbb4081594b2d4fbe61f1719c2b68d6a550e1 — the fp8/MoE cache-clearing was pulled into a shared _clear_refit_memory_caches() helper now called by both offload_before_refit and offload_after_refit, so the two paths can't drift.

File: nemo_rl/models/policy/workers/megatron_policy_worker.py (hunk 7 above)
Comment (terrykong) [reply nemo_rl/models/policy/workers/megatron_policy_worker.py]: fixed in 16ccbb4081594b2d4fbe61f1719c2b68d6a550e1 — collapsed to a single move_model call.

File: nemo_rl/models/policy/__init__.py (hunk 5 above)
Comment (terrykong) [reply nemo_rl/models/policy/__init__.py:549]: fixed in 16ccbb4081594b2d4fbe61f1719c2b68d6a550e1 — both fields now have doc comments, and __init__ raises a fail-fast error at construction if use_coalesced_optimizer_offload is set without use_pinned_optimizer_offload.

File: docs/broken_links_needing_review.json (hunk 3 above)
Comment (terrykong) [reply docs/broken_links_needing_review.json]: fixed in 16ccbb4081594b2d4fbe61f1719c2b68d6a550e1 — docs/broken_links_needing_review.json was removed and the branch rebased onto #2889 (which adds docs/broken_links_false_positives.json).

File: examples/configs/grpo_math_1B.yaml (hunk 4 above)
Comment (terrykong) [reply examples/configs/grpo_math_1B.yaml]: The when-to-enable guidance was added to the inline comments in 16ccbb4081594b2d4fbe61f1719c2b68d6a550e1 — thanks. The other half of this ask (roughly what speedup to expect, since it's a Performance change) still isn't in the PR description; would be great to include some before/after numbers if you have them.

**Review by terrykong — COMMENTED** [review]
@parthmannan can you take a look over this one?

**Comment by terrykong** [comment]
lgtm, but would be good to share any numbers in PR desc to set expectations? also @snivertynv you'll have to rebase
